src/IO/H5/Python/ExtractDatFromH5.py

```diff
@@ -32,10 +40,20 @@ def extract_dat_files(filename, **kwargs):
             raise ValueError(
                 f"Could not make directory '{out_dir}'. Already exists.")
 
+    # We gather all the data first because
+    # 1. The slowest part is actually writing, so that's what we want to
+    #    parallelize (if we are on multiple cores)
+    # 2. So we don't have to worry about accessing an h5file in parallel which
+    #    multiprocessing would have to do if we parsed and wrote in parallel
```

Comment:
Are you sure this actually speeds up the script? The reason I'm asking is that it sounds like the bottleneck here is IO, not the CPU. If an operation is IO bound then parallelizing it to more CPUs is unlikely to speed it up, since the file system is busy with writing files anyway and can't do it any faster (It could actually slow down). I'm confused how the parallelization would help here. Maybe it's the a caching thing (wild guess): switching back and forth between reading the H5 file and writing a text file could be slower than reading everything out of the H5 file first, then writing. I don't know enough about file systems to know for sure. In that case you could just delete the multiprocessing part and get the same speedup. Maybe try that and see if that speeds up the script already?

Then again, loading all dat files into memory before writing them out sounds like it could exceed the RAM, at least in the slow case when the data files are O(GB). Not sure really what to do about this.

Can't we just plot data out of the H5 file? If all you want to do is plot our H5 data with gnuplot (or similar) then take a look here: https://groups.google.com/g/comp.graphics.apps.gnuplot/c/mIUjokGVsSU No need to do expensive extractions.

Reply:
So I tested this on both my local machine and on wheeler. On both machines I was extracting dat files from an H5 file that was about ~4G (they were different files but they were close enough in size). Here is the time it took on each to extract the files:

| Case | Local | Wheeler |
| -- | -- | -- |
| Develop | 4m 37s | 15m 30s |
| This PR on 1 core | 5m 6s | 15m 48s |
| This PR on 8 cores | 1m 14s | 1m 40s |

This PR on 1 core is exactly your suggestion of gathering data first, then writing it on 1 core (no multi-threading happens). So yes I am sure that these changes speed up the script a lot when trying to extract large files, presumably because there's just more cores writing data to different files in parallel.

You make a good point though about RAM usage. What I should probably do is add another option where you can specify which root groups of the H5 file you want to extract. That way you can just exclude the group that has most of the data (\*cough\* cce \*cough\*)

Reply:
If you want to try on wheeler:

```
module load /home/nilsvu/spack/share/spack/modules/linux-rhel7-haswell/h5utils-1.13.1-gcc-9.3.0-xinhocl
h5totxt --help
```

Reply:
I've suspected for a while that the bottleneck for H5 IO is processing done in the hdf5 library rather than actual reading and writing.

Reply:
@wthrowe that sounds plausible but it's not supported by Kyle's findings here I think. He parallelized only the `np.savetxt` part of the script.

Reply:
You're right.  That is weird.

Reply:
So I've tried `h5totxt` and it does work, but it's a bit clunky. The txt files it writes aren't nicely formatted, it can only do one subfile at a time, not a whole group, and in `gnuplot` plotting multiple dat files from multiple h5 files becomes cumbersome. For those reasons, I'd like to keep this utility.

Now it's just a question of what do. I've shown that this parallelization does in fact speed up extraction on a local machine and on a cluster. Are you (@nilsvu) ok with adding this in? I also intend to add a feature where you can only dump certain groups and not others. This will help so we don't even have to worry about large groups. What do you think?

Reply:
Sure, I'm fine with keeping this tool around. To avoid the RAM issue just parallelize the whole read + write routine (see above: https://github.com/sxs-collaboration/spectre/pull/4457#discussion_r1044949107)

Reply:
Ok was able to do this with `h5py`. Our python bindings for some reason can't open read-only h5 files in parallel.

Reply:
Yep we should fix that! Can you open an issue with the error message?